### ISSUE TITLE
forward declarations should be removed

### ISSUE THREAD

**Issue by Laeeth** [atilaneves/dpp#116]
if the .d file declares struct Calendar{ Date date; } and before that there is struct Calendar; then the forward declaration struct Calendar; should not end up in the .d file

**Comment by Laeeth** [comment]
fixed https://github.com/kaleidicforks/dpp/commit/234f99546f290b8ffec47c42ea0e5b7c9db81cd1

**Comment by Laeeth** [comment]
spoke to soon - a bug.

**Comment by atilaneves** [comment]
> we skip forward declarations for aggregates and only want forward declarations for functions/methods

This isn't true - I expect that if you run the tests on your fork they will fail. This code in particular from [this test](https://github.com/atilaneves/dpp/blob/26fc6cf1da87b164c40b462e13b9e4d949fe2de3/tests/it/c/compile/projects.d#L198):

```c
struct Struct;
void fun(struct Struct* s);
struct Struct* make_struct(void);
```

Forward declarations are sometimes necessary. I can't conceive of a way dpp would work on any real life codebase currently if this wasn't currently being handled.

I'm having trouble understanding what the issue is. The C code below correctly gets translated to the D code below it in turn:

```c
struct Foo;
struct Foo { int i; };
```

```d
struct Foo
{
    int i;
}
```

> if the .d file declares struct Calendar{ Date date; } 

Why is the D file be manually declaring C/C++ types?

**Comment by atilaneves** [comment]
> Do you agree that the cases I describe are pathological? Since the file won't compile I don't see how their can be a debate...

I agree there's no debate if dpp can't handle a valid C++ file. I'll have to try running it on xenon myself.

**Comment by atilaneves** [comment]
This issue was in all likelihood caused by how namespaces used to be handled. Now that they're no longer reopened and all cursors are aggregate across namespace declarations with the same name, I doubt this exists. Since there's no concrete example to check against and I haven't run into this in the xenon code I've been running d++ on, I'll close it.

**Comment by Laeeth** [comment]
But it's legal in C to say struct Foo; struct Foo; ad nauseum.  More usually via headers and macros than explicitly.  Since it's legal in C and illegal in D and since DPP is designed to just work unlike dstep, and since no harm can come of removing forward declarations ahead of the definition itself, what harm can come of this?

**Comment by atilaneves** [comment]
> But it's legal in C to say struct Foo; struct Foo; ad nauseum**

Yes.

> what harm can come of this?

I think I wasn't clear - dpp has filtered clang cursors so that there is only ever one declaration since before v0.0.1 - how it would work otherwise on Python, Postgresql, or any other real project?

I'm not saying "dpp shouldn't deal with forward declarations" - I'm saying "dpp already does this". The difference is that when I introduced handling namespaces for C++, the filtering was done differently and ended up having bugs. But only for C++, only for namespaces, and only in certain ways.

Without an actual example in this issue, it was hard for me to understand what was happening until I ran dpp on xenon myself, and have since fixed it in a better way.
